### PR TITLE
Add memory benchmarks

### LINKED ISSUE #107: Add Memory benchmarks
I'm going to try to add memory/size benchmarks to our `cargo bench`.

For size, I'm going to add a simple example binary that will pull ICU4X and we'll measure the impact of that.

### PR DESCRIPTION
**Pull request by gregtatum**
Resolves #107 

Each commit should be logically ordered with descriptions, so I would recommend reviewing on a commit-by-commit basis. I tested this feature using GitHub Actions on my branch of the project. [Here is an example of a CI run with the memory benchmarking.](https://github.com/gregtatum/icu4x/runs/1689595482?check_suite_focus=true)

Clippy seems to be having an issue on my repo, so I'd like to see that here, and maybe to a quick fixup.

## dhat-rs

I decided to go with dhat-rs as the method of instrumentation here as opposed to using just Valgrind for these reasons:

 1. It works on my macOS machine, without me having to load up Linux to get it to work.
 2. We can instrument the memory on each supported platform using the CI.
 3. dhat-rs is on GitHub with an easier contribution model
 4. I have worked with the author in the past in case we need to upstream any changes.
 5. It's written in Rust, which makes it easier to contribute.

## Duckscript

Initially, [I created the a `cargo make` command](https://github.com/unicode-org/icu4x/compare/master...gregtatum:memory-ci-duckscript?expand=1#), but I really struggled with duckscript behaving well. It felt very much like an alpha piece of software. 

Zibi suggested I share my experience. My reasons for having difficulties with it were:

1. Not helpful errors. The line numbers didn't match the yml, and it didn't include examples of surrounding code. The errors were often obtusely written.
2. I struggled with the feature set to accomplish the things I needed to do.
3. I spent a decent amount of time learning the new language.
4. Dealing with argument lists was difficult.
5. There's not much prior art with the language, so it's really hard to search for any actionable solutions to (probably common) problems.
6. Lack of rust-specific features.
7. Lack of tests.

I wrote an injection script that needed AST support, and created a Rust tool to do it. I ended up porting everything in the duckscript command over to this tool (memory_bench). It was a bit more verbose, but I was much happier with the robustness and maintainability of the final solution.

## Benchmark tool

Unfortunately, the [existing benchmark tool](https://github.com/rhysd/github-action-benchmark) does not support a generic data source. I ended up [forking the project](https://github.com/gregtatum/github-action-benchmark/commit/a07750e71f9251c3a5e2da2dab921bd479ff71ba) to add support for `ndjson`, which is JSON delimited by newlines. I plan on attempting to upstream this work, but the repo does not look actively maintained right now. I will update this PR and the code comments with references to the PR once I make it.

## Viewing the benchmark data

* [View the summary of the CI](https://github.com/gregtatum/icu4x/actions/runs/480742710)
* Download the `benchmark-memory` artifact.
* Serve it locally and view the graphs

This should get picked up by the rest of the GitHub actions to be published, however I didn't really know how to verify that.

## Testing plan

I added a few unit tests to the `bench_memory` tool, but not a full end to end test. I believe we have sufficient coverage as this will be running in CI.

## Follow-ups

1. I would like to consolidate the benchmark data under the `benchmarks` folder. Right now the perf benchmarks are under `dev`. I'd like to move it to `benchmarks/perf/*`.
2. Upstream ndjson support.
3. Create links to the newly created charts.

## Discussion

1. How should the alerts be handled? Currently I kept the same defaults as the performance ones, but does this make sense?
2. The following branch must be created before merging: `unmerged-pr-memory-data`
3. Does fail-on-alert for the merge to master step stop the CI from gathering the data points in the benchmark?
4. Should we test on all platforms? It adds a bit more time to our CI. My PR does this as it seemed useful.

/cc @zbraniecki @echeran

### CODE REVIEW

File: .github/workflows/build-test.yml
Comment (sffc) [inline .github/workflows/build-test.yml:301]: ?

File: .github/workflows/build-test.yml
Comment (sffc) [inline .github/workflows/build-test.yml:280]: Is this action pre-existing, or did you create it for icu4x?

File: Cargo.toml
Comment (sffc) [inline Cargo.toml:21]: Discussion point for the group: is this the directory where we want to add this?

File: utils/memory_bench/src/main.rs
Comment (sffc) [inline utils/memory_bench/src/main.rs:73]: Given that you can compile examples irrespective of the component (e.g., `cargo run --example work_log` from the top level package), it would be nice if we can avoid putting the component in the command line argument.  It also makes it harder to scale to new components and examples.

File: utils/memory_bench/src/main.rs
Comment (sffc) [inline utils/memory_bench/src/main.rs:39]: Instead of re-writing the code in this tool, could we add this boilerplate directly to the example files under a feature flag?  For example, `cargo run --features dhat --example work_log`

I think it's okay to carry boilerplate in all of the example files, but if it gets too much, we could put it in a macro that all the examples include.

File: utils/memory_bench/src/main.rs
Comment (sffc) [inline utils/memory_bench/src/main.rs:281]: Why do you need `+nightly`?

File: utils/memory_bench/src/main.rs
Comment (sffc) [inline utils/memory_bench/src/main.rs:280]: See #374; do we need to do anything in this PR to keep the Rust version stable?

File: utils/memory_bench/src/main.rs
Comment (sffc) [inline utils/memory_bench/src/main.rs:306]: Can you interact with dhat more programmatically?  It seems "hacky" to read the results from stderr.

File: .github/workflows/build-test.yml
Comment (gregtatum) [reply .github/workflows/build-test.yml:280]: This is the forked version of the action used by the performance benchmark, it adds support for ndjson.

[https://github.com/gregtatum/github-action-benchmark](https://github.com/gregtatum/github-action-benchmark/commit/a07750e71f9251c3a5e2da2dab921bd479ff71ba)

File: utils/memory_bench/src/main.rs
Comment (gregtatum) [reply utils/memory_bench/src/main.rs:73]: I did this since I was concerned there would be multiple examples named the same thing. I'm not sure what happens then. If we only use example names, then I'm curious if we should change the directory structure as well for the charts.

It would be much simpler in the implementation to iterate over a single list of example names.

File: utils/memory_bench/src/main.rs
Comment (gregtatum) [reply utils/memory_bench/src/main.rs:39]: The instrumentation is not a large burden of code for each example. It's a couple lines of imports, and then a single declaration at the top of the main. I think this is a trade-off between maintaining a tool that rewrites code, and making the examples more complicated for the common case of just running an example. As a user of a project, I would appreciate the examples to be as direct as possible, as I often go there to see how code really works. It would make these examples somewhat less clear to include memory instrumentation.

I would be fine going either way depending on the feedback of the group, but my preference is to include the rewriting tool.

File: .github/workflows/build-test.yml
Comment (gregtatum) [reply .github/workflows/build-test.yml:301]: I'm not sure what the question is exactly here, but I copied the alert settings from the other benchmarks, and added myself to the list. This area is part of the discussion I listed in the main area:

> How should the alerts be handled? Currently I kept the same defaults as the performance ones, but does this make sense?

File: utils/memory_bench/src/main.rs
Comment (gregtatum) [reply utils/memory_bench/src/main.rs:73]: I looked into this again and you can run `cargo run -p icu_datetime --example work_log`. However, if we still want to rewrite the example file, cargo doesn't provide a mechanism to get the path to the example .rs file (as far as I know).

Comment (gregtatum) [reply utils/memory_bench/src/main.rs:73]: You could get the path programmatically by parsing `.toml` files and listing `/examples` dirs, but that's additional complexity.

File: utils/memory_bench/README.md
Comment (gregtatum) [inline utils/memory_bench/README.md:29]: TODO - This is hosted online now by the project maintainer: https://nnethercote.github.io/dh_view/dh_view.html

File: utils/memory_bench/src/main.rs
Comment (sffc) [reply utils/memory_bench/src/main.rs:73]: > I did this since I was concerned there would be multiple examples named the same thing.

I think we can assume this invariant. We should work to prevent example names from conflicting.

> I'm curious if we should change the directory structure as well for the charts.

I'm in favor of treating examples as being mostly flat.

File: utils/memory_bench/src/main.rs
Comment (sffc) [reply utils/memory_bench/src/main.rs:39]: I see rustdoc as the primary place to put code examples.  The example files are a secondary place to put examples, and therefore we can afford a bit of boilerplate in order to make the examples work well with tooling.

We can achieve both goals (making examples useful for testing and simple for users) by having "setup" and "teardown" macros that we include into each example.  Note also that we already have some precedent for changing example files for tooling purposes: in #365 I replaced the built-in Rust main function with `#[no_main]` in order to produce higher quality WASM files for code size testing.

File: utils/memory_bench/src/main.rs
Comment (gregtatum) [reply utils/memory_bench/src/main.rs:306]: I believe the `dhat-heap.json` file could contain this information, if you know how to parse it and read it. It contains an allocation tree summarizing the allocations. The format is a bit dense, but I wrote an importer for the Firefox Profiler trying to understand this data format. It looks like the data might be available if I parsed the JSON, and iterated over it.

Here is the relevant type definition I created for the format in my importer: https://github.com/firefox-devtools/profiler/pull/3128/files#diff-9decaaf66238dad5abffb33cb8ab16232709f9a7b6ce911084489763834aac56R92-R138

Maybe this could be a follow-up? It would require a bit of examination to see if I could pull out the same numbers.

File: utils/memory_bench/src/main.rs
Comment (gregtatum) [reply utils/memory_bench/src/main.rs:281]: The `--profile bench` is an unstable command line flag, which requires the nightly toolchain. I believe I documented this in the CI script, but should probably add a note here. The custom profile is needed for "release" optimizations, but with debug information.

File: utils/memory_bench/src/main.rs
Comment (gregtatum) [reply utils/memory_bench/src/main.rs:280]: This seems like it should be a follow-up tied to #374. I believe we could fix it fairly easily in the CI, but I'm not sure how it works locally.

File: Cargo.toml
Comment (gregtatum) [reply Cargo.toml:21]: In the meeting we decided `tools/memory_bench` instead.

File: utils/memory_bench/src/main.rs
Comment (gregtatum) [reply utils/memory_bench/src/main.rs:39]: In the meeting we decided to add a macro to the examples, and remove the code rewriting.

File: .github/workflows/build-test.yml
Comment (echeran) [inline .github/workflows/build-test.yml:309]: The `@___` syntax in GH Actions specifies a ref. Usually, the refs are tags (named as the semver). If not the semver tag, people usually use the git commit hash. In general, I like versions being immutable, whereas having a branch called `v1` and using it as the ref is not quite that -- can you change `v1` to be the commit hash?

**Review by echeran — COMMENTED** [review]
The GH Actions looks good -- I left just a minor comment.

I thought a little bit more about the meeting's discussion about code injection vs. a more explicit macro usage ("setup"/"teardown"), and I still prefer the latter. Since other reviewers already brought that up and articulated it well, I'll remove myself from the reviewers list by convention so as not to block the PR.

File: utils/memory_bench/src/main.rs
Comment (gregtatum) [reply utils/memory_bench/src/main.rs:73]: It looks like `cargo run` will return an error code if there is an example run from the root with the duplicates, so the CI job should fail. This invariant should be enforced then 👍

> error: `cargo run` can run at most one executable, but multiple were specified

File: Cargo.toml
Comment (gregtatum) [reply Cargo.toml]: I tweaked this slightly to: `tools/benchmark/memory` and `tools/benchmark/macros`

File: utils/memory_bench/src/main.rs
Comment (gregtatum) [reply utils/memory_bench/src/main.rs]: > Given that you can compile examples irrespective of the component (e.g., cargo run --example work_log from the top level package), it would be nice if we can avoid putting the component in the command line argument. It also makes it harder to scale to new components and examples.

So it turns out that features require the path to the Cargo.toml and can't be specified in a virtual workspace. To get around this I ended up specifying a `{package}/{example}` syntax for the tool. I then used the cargo metadata information to locate the Cargo.toml file from the package name.

File: utils/memory_bench/src/main.rs
Comment (gregtatum) [reply utils/memory_bench/src/main.rs]: The new implementation relies on macros and features to enable and disable the memory instrumentation.

File: utils/memory_bench/src/main.rs
Comment (gregtatum) [reply utils/memory_bench/src/main.rs]: To be explicit, I'm not handling this here, and added a note to handle it in #374.

File: utils/memory_bench/src/main.rs
Comment (gregtatum) [reply utils/memory_bench/src/main.rs]: My plan is to follow-up with this one, rather than handle it in the initial PR. https://github.com/unicode-org/icu4x/issues/460

File: .github/workflows/build-test.yml
Comment (gregtatum) [reply .github/workflows/build-test.yml]: Yes, that sounds sensible to me, and I handled it. We should probably do this as well for `rhysd/github-action-benchmark`. This is the recommended practice for hardening security:

https://docs.github.com/en/actions/learn-github-actions/security-hardening-for-github-actions#using-third-party-actions

One thing on this new approach is that I'm getting a warning notice which is a bit confusing.

From: https://github.com/gregtatum/icu4x/actions/runs/501955931

> gregtatum/github-action-benchmark@d3f06f738e9612988d575db23fae5ca0008d3d12 looks like the shortened version of a commit SHA. Referencing actions by the short SHA will be disabled soon. Please see https://docs.github.com/en/actions/learn-github-actions/security-hardening-for-github-actions#using-third-party-actions. 

However, `d3f06f738e9612988d575db23fae5ca0008d3d12` is the full hash as far as I can tell.

**Comment by gregtatum** [comment]
This is ready for re-review. All feedback should be addressed. This did involve significant rewriting of the patch, so a lot of the code will be different.

Summary of major changes:
* Consolidated to `/tools/benchmark` folder
* Removed code rewriting
* Added macro support
* Memory benchmarking is enabled with a feature.
* Fewer `matrix` job options, it's only per-platform now. Each job runs all of the examples.
* Introduction of the `{package}/{example}` syntax for invoking the memory instrumentation on an example.

Here is a full run in CI:

https://github.com/gregtatum/icu4x/actions/runs/501955931

File: components/datetime/examples/work_log.rs
Comment (sffc) [inline components/datetime/examples/work_log.rs:10]: Optional: here and elsewhere: if you only use the macro one time, I have a slight preference for

```
icu_benchmark_macros::static_setup!()
```

and down below

```
icu_benchmark_macros::main_setup!()
```

and then remove the `use` line.

File: components/datetime/examples/work_log.rs
Comment (zbraniecki) [inline components/datetime/examples/work_log.rs:40]: nit; could you instead not use `use *;` and do:

```rust
icu_benchmark_macros::static_setup!();
```
```rust
icu_benchmark_macros::main_setup!();
```

This may seem like more boilerplate, but it makes it very visually clean that those two lines are benchmark related macros and can be skipped when interpreting the code.

File: tools/benchmark/README.md
Comment (zbraniecki) [inline tools/benchmark/README.md:36]: could you document here how to use firefox profiler with that?

**Review by zbraniecki — APPROVED** [review]

**Review by sffc — DISMISSED** [review]
This is much better; thank you!  Just one suggestion

File: components/datetime/examples/work_log.rs
Comment (gregtatum) [reply components/datetime/examples/work_log.rs:10]: I agree, this is nicer.

File: components/datetime/examples/work_log.rs
Comment (gregtatum) [reply components/datetime/examples/work_log.rs:40]: I agree, this is nicer.

File: tools/benchmark/README.md
Comment (gregtatum) [reply tools/benchmark/README.md:36]: Sure, it's still pretty experimental, but probably useful. I'll note it as such.

**Comment by gregtatum** [comment]
I addressed the last few comments. Someone will need to create the `unmerged-pr-memory-data` branch before merging. I don't have access to do so.

**Review by sffc — DISMISSED** [review]

**Review by zbraniecki — DISMISSED** [review]

**Review by nciric — APPROVED** [review]

**Review by sffc — APPROVED** [review]

**Review by echeran — APPROVED** [review]

**Pull request merged**
